sw/device/lib/base/status.h

```diff
@@ -0,0 +1,135 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_STATUS_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_STATUS_H_
+#include <assert.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+
+#define USING_ABSL_STATUS
+#include "sw/device/lib/base/absl_status.h"
+#undef USING_ABSL_STATUS
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * We use the error category codes from absl_status.h.  We build a packed
+ * status value that identifies the source of the error (in the form of the
+ * module identifier and line number).
+ *
+ * By default, the module identifier is the first three letters of the
+ * source filename.  The identifier can be overridden (per-module) with the
+ * DECLARE_MODULE_ID macro.
+ *
+ * Our status codes are arranged as a packed bitfield, with the sign
+ * bit signifying whether the value represents a result or an error.
+ *
+ * All Ok (good) values:
+ * 32  31                                             0
+ *  +---+---------------------------------------------+
+ *  |   |                  31 bit                     |
+ *  | 0 |                  Result                     |
+ *  +---+---------------------------------------------+
+ *
+ * All Error values:
+ * 32  31      26      21      16             5       0
+ *  +---+-------+-------+-------+-------------+-------+
+ *  |   |   15 bit              | 11 bit      | 5 bit |
+ *  | 1 |   Module Identifier   | Line Number | code  |
+ *  +---+-------+-------+-------+-------------+-------+
+ *
+ * The module identifier value is interpreted as three 5-bit fields
+ * representing the characters [0x40..0x5F] (e.g. [@ABC ... _]).
+ */
+typedef union Status_ {
+  int32_t value;
+  struct {
+    uint32_t code : 5;
+    uint32_t arg : 11;
+    uint32_t module_id : 15;
+    uint32_t is_error : 1;
+  } error;
```

Comment:
Two concerns:
1- Unions can be tricky (you are not supposed to read from an inactive member)
2- I never encountered an issue personally but heard a lot of things about bitfields and reproducibility of the layout across platforms and compilers.

Just curious if there is a particular reason for not using something like:
```
typedef struct status {
   uint32_t code;
} status_t;
```
along with the necessary `bitfield_t` definitions?

Reply:
The reasons for this are (a) old habits.  Type punning via unions is a classic technique and (b) I find the `bitfield_*` functions extremely distasteful.  Nevertheless, I've changed this around to use those functions.